src/accelerate/data_loader.py

```diff
@@ -539,6 +553,7 @@ def __iter__(self):
                 current_batch = next_batch
             except StopIteration:
                 self.end_of_dataloader = True
+                self._update_state_dict()
```

Comment:
additionally, we need to update the state dict when we've hit the stop iteration to know if we've hit the end of the iterator